core/Tracker/Failures.php

```diff
@@ -0,0 +1,178 @@
+    }
+
+    public function getFailuresForSites($idSites)
+    {
+        if (empty($idSites)) {
+            return array();
+        }
+        $idSites = array_map('intval', $idSites);
+        $idSites = implode(',', $idSites);
+        $failures = PiwikDb::fetchAll(sprintf('SELECT * FROM %s WHERE idsite IN (%s)', $this->tablePrefixed, $idSites));
+        return $this->enrichFailures($failures);
+    }
+
+    public function deleteTrackingFailure($idSite, $idFailure)
+    {
+        PiwikDb::query(sprintf('DELETE FROM %s WHERE idsite = ? and idfailure = ?', $this->tablePrefixed), array($idSite, $idFailure));
+    }
+
+    public function deleteTrackingFailures($idSites)
+    {
+        if (!empty($idSites)) {
+            $idSites = array_map('intval', $idSites);
+            $idSites = implode(',', $idSites);
+            PiwikDb::query(sprintf('DELETE FROM %s WHERE idsite IN(%s)', $this->tablePrefixed, $idSites));
+        }
+    }
+
+    public function deleteAllTrackingFailures()
+    {
+        PiwikDb::query(sprintf('DELETE FROM %s', $this->tablePrefixed));
+    }
+
+    private function enrichFailures($failures)
+    {
+        foreach ($failures as &$failure) {
+            try {
+                $failure['site_name'] = Site::getNameFor($failure['idsite']);
+            } catch (UnexpectedWebsiteFoundException $e) {
+                $failure['site_name'] = Piwik::translate('General_Unknown');
+            }
+            $failure['pretty_date_first_occurred'] = Date::factory($failure['date_first_occurred'])->getLocalized(Date::DATETIME_FORMAT_SHORT);
+            parse_str($failure['request_url'], $params);
+            if (empty($params['url'])) {
+                $params['url'] = ' ';// workaround it using the default provider in request constructor
+            }
+            $request = new Request($params);
+            $failure['url'] = trim($request->getParam('url'));
+            $failure['problem'] = '';
+            $failure['solution'] = '';
+            $failure['solution_url'] = '';
+
+            switch ($failure['idfailure']) {
+                case self::FAILURE_ID_INVALID_SITE:
+                    $failure['problem'] = Piwik::translate('CoreAdminHome_TrackingFailureInvalidSiteProblem');
+                    $failure['solution'] = Piwik::translate('CoreAdminHome_TrackingFailureInvalidSiteSolution');
+                    $failure['solution_url'] = 'https://matomo.org/faq/todo...';
+                    break;
+                case self::FAILURE_ID_NOT_AUTHENTICATED:
+                    $failure['problem'] = Piwik::translate('CoreAdminHome_TrackingFailureAuthenticationProblem');
+                    $failure['solution'] = Piwik::translate('CoreAdminHome_TrackingFailureAuthenticationSolution');
+                    $failure['solution_url'] = 'https://matomo.org/faq/todo...';
+                    break;
+            }
+        }
+
+        Piwik::postEvent('Tracking.makeFailuresHumanReadable', array(&$failures));
```

Comment:
Should this have an @ignore or docs?

Reply:
yes 👍  will add an ignore for now, and add proper docs as soon as we use it in another plugin. don't want to make it an API just yet before not having actually verified it works well like this.